ledger-service/jwt/src/main/scala/com/digitalasset/jwt/JwksVerifier.scala

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.jwt
+
+import java.net.{URI, URL}
+import java.security.interfaces.RSAPublicKey
+import java.util.concurrent.TimeUnit
+
+import com.auth0.jwk.UrlJwkProvider
+import com.digitalasset.jwt.JwtVerifier.Error
+import com.google.common.cache.{Cache, CacheBuilder}
+import scalaz.{-\/, Show, \/}
+import scalaz.syntax.show._
+
+/**
+  * A JWK verifier, where the public keys are automatically fetched from the given JWKS URL.
+  *
+  */
+class JwksVerifier(url: URL) extends JwtVerifierBase {
+  private[this] val http = new UrlJwkProvider(url)
+
+  private[this] val cache: Cache[String, JwtVerifier] = CacheBuilder
+    .newBuilder()
+    .maximumSize(10)
+    .expireAfterWrite(10, TimeUnit.HOURS)
```

Comment:
10 hr expiration timeout... isn't it too long? If key is revoked, we would still allow tokens signed by this key until the cache expires.

Reply:
Good point.

If you are worried about revocation, you can generate JWT tokens with a short expiration time (and check the validity of the key pair every time you sign a token).

Given that, do we still need a short expiration time on the key cache? Note that on a cache miss, the key has to be fetched from a remote URL in a blocking call, which will significantly increase the time to process the corresponding gRPC request.

Reply:
Yep, that will do it.
However I thought you need to create a verifier with one of `acceptLeeway`, `acceptNotBefore` or `acceptExpiresAt` in order to have it checking the expiration.

Reply:
in any case it would be nice if we can have this timeout configurable via an explicit argument on the `JwksVerifier` (without default value), this way we will force our users to think about this timeout. We should not be picking it for them.

Reply:
Not sure keys ever really "expire" in a JWKS context…it's the tokens that are signed with those keys that expire. The only compelling reason I'd see to invalidate this cache is because of concerns on unmitigated memory growth, but since you set a `maximumSize`, that's probably good enough.

It should be considered a developer error to _ever_ reuse a key for the same `kid` (and also, terribly, our keys for Auth0 seem to have never been changed so far).

Reply:
a key can be revoked, right?

Reply:
I agree with @da-tanabe that handling revocation of access rights should be done through tokens with a short expiration time. 

Moving the values to the constructor sounds like a good idea though.

The expiration values for the cache entries are copied from https://github.com/auth0/jwks-rsa-java/blob/master/src/main/java/com/auth0/jwk/GuavaCachedJwkProvider.java#L28. This class is very similar to that, except that we cache `Verfiers` and not just the `JWK` data. Also, it will allow us to better log authentication related events.

Reply:
:+1: